mantle/platform/qemu.go

```diff
@@ -375,17 +336,32 @@ func (builder *QemuBuilder) ConsoleToFile(path string) {
 	builder.Append("-display", "none", "-chardev", "file,id=log,path="+path, "-serial", "chardev:log")
 }
 
-func (builder *QemuBuilder) EnableUsermodeNetworking(forwardedPort uint) {
+func (builder *QemuBuilder) EnableUsermodeNetworking(h []HostForwardPort) {
+	builder.UsermodeNetworking = true
+	builder.requestedHostForwardPorts = h
+}
+
+func (builder *QemuBuilder) usermodeNetworkingAssignPorts() error {
 	netdev := "user,id=eth0"
-	if forwardedPort != 0 {
-		netdev += fmt.Sprintf(",hostfwd=tcp:127.0.0.1:0-:%d", forwardedPort)
+	for i := range builder.requestedHostForwardPorts {
+		address := fmt.Sprintf(":%d", builder.requestedHostForwardPorts[i].HostPort)
+		l, err := net.Listen("tcp", address)
```

Comment:
When host port 0 is specified, the system will find a suitable available port to use.  The `Listener` is the trigger for the system to do this.  Ideally all callers of this function will use port 0 so there are no port conflicts. A caller can specify a port but it will fail if the port is already in use.  

I thought about failing over to using an available port if the caller specifies a port that is in use, but that may be confusing since they explicitly set the port.

Reply:
It's still a bit racy between the listen/close then launching qemu, something else could grab the port.  But that's hard to fix without passing the socket to qemu.  I don't think that's possible at the moment though.

Reply:
@cgwalters  I can add a comment about possible race condition here.  Doing it this way just made the port management much simpler.

Reply:
Oh yeah, adding a comment is fine, I was just noting this in passing.

Reply:
Added comment